### PR TITLE
feat: add custom commands for piping and unpiping text

### PR DESCRIPTION
**Pull request by polvalente**
This PR aims to bring the functionality developed [here](https://github.com/polvalente/ex-plumber/pull/1) to ElixirLS as a custom language server command, as suggested by @axelson.

~As of opening of this PR, this is still a work in progress so implementation details can be discussed~

### CODE REVIEW

**Comment by polvalente** [comment]
> I'm excited for this change. I think it will be a helpful refactoring to support.
> 
> In terms of code structure can you extract a "pure" functional module from `ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes` to handle the core piping and unpiping logic? I guess this might result in an even longer name which is a bit awkward but is probably fine for now. The separate module will make adding more test cases easier because they'll require less setup.
> 
> Also I believe it should be possible to implement this without a regex scan, which should help the resiliency of this code. Maybe @lukaszsamson has some specific ideas on that front.

Sure, I'll split the module further. Should I leave that as a submodule of the command?

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            {remaining_text,
+             %{
+               acc
+               | walked_text: acc.walked_text <> current_char,
+                 function_call: function_call,
+                 range: call_range
+             }}
+          else
+            {remaining_text, %{acc | walked_text: acc.walked_text <> current_char}}
+          end
```
Comment (yashin5) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:72]:
```suggestion
        fn current_char, remaining_text, ^line, ^col, acc ->
            {:ok, function_call, call_range} =
              get_function_call(line, col, acc.walked_text, current_char, remaining_text)

            {remaining_text,
             %{
               acc
               | walked_text: acc.walked_text <> current_char,
                 function_call: function_call,
                 range: call_range
             }}
             current_char, remaining_text, _current_line, _current_col, acc -> 
            {remaining_text, %{acc | walked_text: acc.walked_text <> current_char}}
          end
```

**Comment by polvalente** [comment]
@axelson I've split the module into one which handles the AST formatting and another which handles the string parsing and the command interface. Should I also move the string parsing to a separate module? IMO it's good to have a kind of "integration test" so I wouldn't remove any tests anyway.

I've also managed to refactor the code so Regex.scan could be removed

File: apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs [hunk 2]
```diff
@@ -0,0 +1,159 @@
+defmodule ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes.ASTTest do
```
Comment (polvalente) [inline apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs:1]: Most of these test I imported from my original repo, so if any of these tests feel redundant, feel free to make suggestions for me to remove them!

**Comment by polvalente** [comment]
Finally, there are some tests failing in earlier versions of Elixir but still yielding results that make sense. Should I include an `assert X or Y` in those, or maybe make the assertion depend on the Elixir version somehow?

File: apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+    test "multi-line selection with two args in named function" do
+      assert "X.Y.Z.a |> X.Y.Z.function_name(b, c)" ==
+               AST.to_pipe("""
+                 X.Y.Z.function_name(
+                 X.Y.Z.a,
+                 b,
+                 c
+               )
+               """)
```
Comment (axelson) [inline apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs:32]:
```suggestion          
      actual =
        AST.to_pipe("""
        X.Y.Z.function_name(
          X.Y.Z.a,
          b,
          c
          )
        """)

      expected =
        if Version.match?(System.version(), "< 1.10.0") do
          "X.Y.Z.a() |> X.Y.Z.function_name(b, c)"
        else
          "X.Y.Z.a |> X.Y.Z.function_name(b, c)"
        end

      assert actual == expected
```
We should change the Elixir version-dependant tests to use different assertions for tests that vary based on the Elixir version. But in general we should strive to not have tests that are specific to an Elixir version. I need to look at the rest of the code but could we generate `X.Y.Z.a()` version for all elixir versions? I feel that might be slightly more idiomatic (although since the formatter will accept either `X.Y.Z.a()` or `X.Y.Z.a` it is a little ambiguous.

File: apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs [hunk 4]
```diff
@@ -0,0 +1,159 @@
+defmodule ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes.ASTTest do
+  use ExUnit.Case
+
+  alias ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes.AST
+
+  describe "to_pipe/1" do
+    test "single-line selection with two args in named function" do
+      assert "A.B.C.a |> X.Y.Z.function_name(b)" == AST.to_pipe("X.Y.Z.function_name(A.B.C.a, b)")
```
Comment (axelson) [inline apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs:8]:
```suggestion
      assert AST.to_pipe("X.Y.Z.function_name(A.B.C.a, b)") == "A.B.C.a |> X.Y.Z.function_name(b)"
```

For these tests can we put the expected result on the right side? Roughly we follow https://github.com/christopheradams/elixir_style_guide#testing

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+  @behaviour ElixirLS.LanguageServer.Providers.ExecuteCommand
+
+  @impl ElixirLS.LanguageServer.Providers.ExecuteCommand
+  def execute(
+        %{"uri" => uri, "cursor_line" => line, "cursor_column" => col, "operation" => operation},
+        state
+      )
+      when is_integer(line) and is_integer(col) and is_binary(uri) and
+             operation in ["to_pipe", "from_pipe"] do
+    # line and col are assumed to be 1-indexed
```
Comment (axelson) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:23]: While a bit awkward, for consistency we should assume these are both 0-indexed because that's the standard that LSP uses (as well as the expandMacro command):
> A position inside a document (see Position definition below) is expressed as a zero-based line and character offset. The offsets are based on a UTF-16 string representation.

https://microsoft.github.io/language-server-protocol/specification#textDocuments

**Review by axelson — COMMENTED** [review]
> Finally, there are some tests failing in earlier versions of Elixir but still yielding results that make sense. Should I include an assert X or Y in those, or maybe make the assertion depend on the Elixir version somehow?

I put some details in a comment, but I think we should change the expected result based on the elixir version

Running out of time for this review, but I wanted to send what I have. Will take another look in the future.

File: apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs (hunk 3 above)
Comment (polvalente) [reply apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs:32]: I think that changing the tests cases for having `X.Y.Z.a()` as the argument would fix that.
Either that, or forcing parens or no parens throught the AST module would remove the need for making the tests version-dependent

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex (hunk 5 above)
Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:23]: No problem. As I didn't have clear standard to follow, I assumed 1 indexed because VSCode shows them as 1-indexed in the  bottom toolbar. However, changing this I think is just a matter of subtracting 1 somewhere, with no changes to core functionality

Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex]: @axelson I've implemented both this change and the test assertion order. Hopefully it's ready to go now!

Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex]: Nevermind, I still have to fix the tests which are breaking due to versioning. I'll add the parens to the original AST so we don't have to deal with this minor detail

Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex]: All tests pass now :D

File: apps/language_server/.tool-versions [hunk 6]
```diff
@@ -0,0 +1 @@
+elixir 1.11.3-otp-23
```
Comment (lukaszsamson) [inline apps/language_server/.tool-versions:1]: some time ago we resigned from using .tool-versions

Comment (polvalente) [reply apps/language_server/.tool-versions:1]: Oh, this shouldn't have been commited! I ended up setting local versions to fix a broken test between < 1.10 and >= 1.10

I'll revert this.

Comment (polvalente) [reply apps/language_server/.tool-versions]: Done :)

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+    {head, _tail} = String.split_at(call, -String.length(tail))
+    col = col - String.length(head) + 1
+
+    {:ok, call, range(line, col, end_line, end_col)}
+  end
+
+  defp do_get_function_call(text, start_char, end_char) do
+    text
+    |> String.graphemes()
```
Comment (lukaszsamson) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:148]: This could be converted to recursive binary matching

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+        {:cont, [c | acc]}
+      end
+    end)
+    |> IO.iodata_to_binary()
+  end
+
+  defp count_newlines_and_get_tail(s) do
+    for <<c::binary-size(1) <- s>>, reduce: {0, 0} do
+      {count, tail} ->
+        if c == "\n" do
```
Comment (lukaszsamson) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:267]: Add /r/n handling

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+        {:ok, %{edited_text: unpiped_text, edit_range: range}}
+    end
+  end
+
+  defp get_function_call(line, col, head, current, original_tail) do
+    tail = do_get_function_call(original_tail, "(", ")")
+
+    {end_line, end_col} =
+      if String.contains?(tail, "\n") do
+        tail_list = String.split(tail, "\n")
```
Comment (lukaszsamson) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:128]:
```suggestion
        tail_list = String.split(tail, "\r\n", "\n", "\r")
```

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+
+    {line_offset, tail_length} = pipe_left |> String.reverse() |> count_newlines_and_get_tail()
+
+    start_line = line - line_offset
+
+    start_col =
+      if line_offset != 0 do
+        head
+        |> String.trim_trailing(pipe_left)
+        |> String.split("\n")
```
Comment (lukaszsamson) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:214]:
```suggestion
        |> String.split("\r\n", "\n", "\r")
```

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex (hunk 8 above)
Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:267]: @lukaszsamson doesn't this already cover that case? The main thing here is to get the number of lines and any remaining characters after "\n"

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex (hunk 9 above)
Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:128]: Do "\r" characters alone imply newlines? I though they just implied "go back to the first column"

Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:128]: nvm, for Mac OS before OS X they do imply newlines

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex (hunk 8 above)
Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:267]: Also disregard this. For covering "\r" some refactoring here is needed

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex [hunk 11]
```diff
@@ -0,0 +1,59 @@
+defmodule ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes.AST do
```
Comment (lukaszsamson) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex:1]: I'd add some conditions to matching expressions in this module. We can filter some false positives if we validate that in `{call, _, args} when call not in [:., :__aliases__, :"::", :{}, :|>]`

Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex:1]: Nice. Could you give some examples which break over the current code, so I can add them as test cases?

File: apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+               X.Y.Z.a(),
+               b,
+               c
+             )
+             """) == "X.Y.Z.a() |> X.Y.Z.function_name(b, c)"
+    end
+  end
+
+  describe "from_pipe/1 single-line" do
+    test "three args in named function" do
```
Comment (lukaszsamson) [inline apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs:37]: Add test where there is no pipe operator

File: apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs [hunk 13]
```diff
@@ -0,0 +1,159 @@
+defmodule ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes.ASTTest do
+  use ExUnit.Case
+
+  alias ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes.AST
+
+  describe "to_pipe/1" do
+    test "single-line selection with two args in named function" do
```
Comment (lukaszsamson) [inline apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs:7]: Add some tests on negative examples:
- to_pipe on 0 arity function call
- to_pipe on atom
- to_pipe on calls with no parens, e.g. `X.fun`, `Y.fun "abc"`
- already a piped call, e.g. `x |> Y.fun`

File: apps/language_server/test/providers/execute_command/manipulate_pipes_test.exs [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+      end
+    end
+  end
+
+  describe "execute/2 to_pipe" do
+    test "can pipe remote calls in single lines" do
+      {:ok, _} =
+        JsonRpcMock.start_link(success_reply: {:ok, %{"applied" => true}}, test_pid: self())
+
+      uri = "file:/some_file.ex"
```
Comment (lukaszsamson) [inline apps/language_server/test/providers/execute_command/manipulate_pipes_test.exs:36]: this is an invalid unix file URI
```suggestion
      uri = "file:///some_file.ex"
```

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+  def execute(
+        %{"uri" => uri, "cursor_line" => line, "cursor_column" => col, "operation" => operation},
+        state
+      )
+      when is_integer(line) and is_integer(col) and is_binary(uri) and
+             operation in ["to_pipe", "from_pipe"] do
+    # line and col are assumed to be 0-indexed
+    source_file = Server.get_source_file(state, uri)
+
+    {:ok, %{edited_text: edited_text, edit_range: edit_range}} =
```
Comment (lukaszsamson) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:26]: Since you are using the column parameter to get the correct ranges you need to convert it from UTF16 to UTF8 index. We do this e.g. in https://github.com/elixir-lsp/elixir-ls/blob/794ed49a571f7943d478ef0a87ed94ab07b30c67/apps/language_server/lib/language_server/source_file.ex#L199

File: apps/language_server/test/providers/execute_command/manipulate_pipes_test.exs [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+
+      text = """
+      defmodule A do
```
Comment (lukaszsamson) [inline apps/language_server/test/providers/execute_command/manipulate_pipes_test.exs:39]: Add one to_pipe and one from_pipe integration test with \r\n line endings

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex (hunk 11 above)
Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex:1]: One of your suggested cases needed this to work :)

File: apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs (hunk 13 above)
Comment (polvalente) [reply apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs:7]: The current code doesn't pipe this correctly. It pipes `a + b` to `a |> +b`, which raises `(ArgumentError) piping into a unary operator is not supported`

Maybe we could add infix operators to a blocklist of pipeable functions, like you did with `:__aliases__` and so on. How does that sound?

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex (hunk 11 above)
Comment (lukaszsamson) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex:1]: I don't have examples off the top of my head but we use this check in elixir_sense https://github.com/elixir-lsp/elixir_sense/blob/7fb31589e9dfdb91c6498e1169a13a0fe056d15b/lib/elixir_sense/core/metadata_builder.ex#L30

File: apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs (hunk 13 above)
Comment (lukaszsamson) [reply apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs:7]: > Maybe we could add infix operators to a blocklist of pipeable functions, like you did with :__aliases__ and so on. How does that sound?

If that's easier to do then I'm ok with that. It's better to fail fast than break user's code.

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex (hunk 15 above)
Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:28]: @lukaszsamson I've applied all requested changes. I've looked at this code, but I don't completely follow how to apply it to my PR. Could you perhaps add a suggestion to where I should add this?

I tried adding this to the beggining of the pipeline, but the cursor stopped pointing at the correct character and thus the code broke

File: apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs (hunk 13 above)
Comment (polvalente) [reply apps/language_server/test/providers/execute_command/manipulate_pipes/ast_test.exs:7]: Treating this correctly would perhaps be impossible. For piping you need the fully qualified name, but due to the way we parse things, we can't actually resolve the module name, given that infix operators are always imported from somewhere.

I'll add a check for operators to be skipped

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex (hunk 8 above)
Comment (lukaszsamson) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex]: \r\n is a bit tricky to do right. LSP treats \n, \r and \n\r as newlines but elixir parser only supports \n and \r\n. On the other hand elixir allows \r inside strings, charlists and heredocs. I once needed something similar in one of the tests https://github.com/elixir-lsp/elixir-ls/blob/794ed49a571f7943d478ef0a87ed94ab07b30c67/apps/language_server/test/source_file_test.exs#L67

Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex]: Yeah, these are showing to be the hardest part of the code haha

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+  @moduledoc """
+  AST manipulation helpers for the `ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes`\
+  command.
+  """
+
+  @doc "Parses a string and converts the first function call, pre-order depth-first, into a pipe."
+  def to_pipe(code_string) do
+    {piped_ast, _} =
+      code_string
+      |> Code.string_to_quoted!()
```
Comment (polvalente) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex:11]: This part of the code breaks upon `\r` characters. I had added a String.replace here to deal with that, but that broke literal strings.

Since we can have arbitrary sigils as well, I'm starting to think that we'll have to just not support `\r` as a line separator.
What do you think about this @lukaszsamson?

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex (hunk 15 above)
Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:28]: I've added a test that breaks on 2-byte characters for the "from_pipe" command, so there's that.
However, I'm still kinda lost on how to deal with the initial offsetting and how to convert the output ranges correctly so I can work with purely utf-16 binaries or something like that

Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:28]: One thing I looked into is that ElixirSense's walk_text uses String.next_grapheme to recurse through the binary, and that ends up dealing with multi-byte characters.

```elixir
iex(8)> "olá\r\nç\nç" |> String.graphemes                                                                      
["o", "l", "á", "\r\n", "ç", "\n", "ç"]
```

I think that if I solve the issue on my custom recursions the code will yield the expected ranges. Any thoughts on this?

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex (hunk 5 above)
Comment (axelson) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex]: Great, thanks!

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex (hunk 15 above)
Comment (lukaszsamson) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:28]: Yes that part of LSP/Elixir integration is really messy. To give you an example suppose you have range
```
{"line": 1, character: 2} to {line: 6, character: 5}
```
Character indices are codepoints in UTF16 encoding but elixir String are in UTF8 (and graphemes are a different beast as you noticed)
To get the source slice range right you need to
1. split it into lines
2. convert first (1) and last (6) line into UTF16
3. take slice [2 * 16..] from first line and convert it back into UTF8
4. append lines 2-5 (no need to convert here)
5. take slice [0..5 * 16] from last line and convert it back into UTF8

But let's leave it for another PR. There are more places in the codebase where it's not addressed properly.

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex (hunk 17 above)
Comment (lukaszsamson) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes/ast.ex:11]: Good. Let's add a quick check fail fast when \r is detected.

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+  defp do_get_pipe_call(text, acc \\ {"", false, false})
+
+  defp do_get_pipe_call(_text, {acc, _, true}), do: acc
+  defp do_get_pipe_call("", {acc, _, _}), do: acc
+
+  defp do_get_pipe_call(<<?\r::utf16, ?\n::utf16, _::bitstring>>, {acc, true, _}),
+    do: <<?\r::utf16, ?\n::utf16, acc::bitstring>>
+
+  defp do_get_pipe_call(<<0, c::utf8, _::bitstring>>, {acc, true, _})
+       when c in [?\t, ?\v, ?\r, ?\n, ?\s],
```
Comment (lukaszsamson) [inline apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:249]: is anyone still using vertical tabs?

**Review by lukaszsamson — APPROVED** [review]

**Comment by lukaszsamson** [comment]
Great job @polvalente, it's going to be a nice addition.

**Comment by polvalente** [comment]
> Great job @polvalente, it's going to be a nice addition.

Thabk you! I learned a lot with this PR and the reviews were great!

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex (hunk 18 above)
Comment (polvalente) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:249]: I've never encountered one in the wild haha

File: apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex (hunk 15 above)
Comment (lukaszsamson) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:28]: @victorolinasc
The reason we switch to latin1 is to use binary mode on stdio instead of the default character mode (which is charset dependant). We just send and receive bytes on the wire. The LSP protocol messages are UTF8 encoded as the spec requires but thats on another layer (see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#baseProtocol). The indices used in LSP requests and responses use character counts in UTF16 (see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocuments) but the files themselves are not transferred over LSP and read directly by language server. Thus we end up with UTF8 encoded elixir binaries that are indexed by UTF16 character counts. Everything works fine as long as we stay inside ASCII range.

Comment (axelson) [reply apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex:28]: Some other relevant links: 
* feature request for LSP to support UTF-8: https://github.com/microsoft/language-server-protocol/issues/376
* UTF-8 extension supported by rust-analyzer and clangd: https://clangd.llvm.org/extensions.html#utf-8-offsets

**Review by axelson — APPROVED** [review]
Looks good :raised_hands:!

Excited to try this out :heart:

Now it would be good to add this command to vscode-elixir-ls (PR will probably look similar to https://github.com/elixir-lsp/vscode-elixir-ls/pull/176)

**Comment by polvalente** [comment]
> Looks good !
> 
> Excited to try this out 
> 
> Now it would be good to add this command to vscode-elixir-ls (PR will probably look similar to [elixir-lsp/vscode-elixir-ls#176](https://github.com/elixir-lsp/vscode-elixir-ls/pull/176))

@axelson If no one tackles this before me, I might be able to open the pull request later in the weekend!
